### DISCUSSION TITLE
프로젝트 개발 전 논의 사항

### DISCUSSION THREAD

**Discussion by belljun3395** [YAPP-Github/24th-Web-Team-1-BE#3] [General]
# 프로젝트 개발 전 논의 사안 정리

### 역할 분배

- [x] 플젝 초기 셋업 @hun-ca 

- [x] CI/CD @belljun3395  / @hun-ca 에게 컨펌 -> #9 

### 인프라

- NCP를 사용한다
- ~~NCP의 DEV TOOL을 사용해 본다.~~
- CD의 경우 클라우드사와 연관되지 않도록 구성한다.
  - 깃헙 액션 + SSH
    - 무중단 배포를 위해 Nginx 도입
- 테라폼을 도입한다.
- 배치 관련된 것은 우선 따로 서버를 두지 않는다.
  - 깃헙 액션 + 테라폼으로 배치 서버를 일시적으로 사용하는 것이 가능하다면 도입 시도
  - 배치는 우선 API로 구현한다.
  - 배치 API 동작은 깃헙 액션 크론을 통해 구현한다.

### DB

- RDB를 사용한다면 FK를 안잡는다.
  - 엔티티 관계 표시를 위해 문서를 잘 적성하거나 클래스 상속을 사용한다.

### 소스 코드

- 도메인 모델을 도입한다 -> 엔티티에는 메서드가 없다.
- UseCase#execute 패턴을 도입한다.
- 조회 기능에는 Query 객체를 사용한다.

### 모듈 

- common

- **api** (common, data, email )
- data
- email

- **batch** ( email )
 
### API 명세

- 컨트롤러 테스트를 통해 Swagger Docs를 생성한다.
- Swagger를 활용해서 Admin 페이지를 제공

### 클라 협의

- 클라 배포는 클라에서 cloudflare로
- 이메일 전송은 서버, 템플릿은 클라에서

### 언어

- [ ] 코루틴, webflux 공부하고 결정하기

### 인프라

- [x] 구체적 지원 항목 확인 필요

### DB

- 어떤 DB를 사용할지 미정
- 1,2차 MVP에 쓰기 기능이 존재하지 않는다  -> MonogoDB, 존재한다 -> MySQL + Redis
- RDB를 사용할시 JPA 밴, Jooq(navtive query) + Flyway 활용 고려

### 소스코드

- [ ] 예제 코드 작성하며 맞춰 나간다. @hun-ca , @belljun3395 

### 깃 컨밴션

- [x] @hun-ca 가 평소에 사용하던 방법 공유

**Comment by belljun3395** [comment]
# 인프라 논의 사항

AWS, NCP 선택 필요

AWS: 익숙하다, 레퍼런스가 많다

NCP: 비용을 지원해준다.

**Reply by belljun3395** [reply]
개인적으로 어떤 클라우드를 사용하던 **terraform**을 사용해보고 싶습니다!
코드로 인프라를 관리하다 보니까 서로 인프라 관련된 공유나 학습에 관해서도 보다 편할 것 같아요!

**Reply by belljun3395** [reply]
추가로 알아보니까 깃헙 액션도 [REST API](https://docs.github.com/ko/rest/actions/workflows?apiVersion=2022-11-28#create-a-workflow-dispatch-event)를 지원하더라고요
액션 안에 잡만 잘 작성하면 이메일 서버나 배치 서버 같은거 잠시 고사양으로 만들고 삭제하는 방식으로도 사용할 수 있을 것 같아요!

EX)
1.  terraform으로 자원 생성
2.  생성된 자원에서 필요한 기능 수행
3.  terraform으로 자원 삭제

**Reply by hun-ca** [reply]
> 개인적으로 어떤 클라우드를 사용하던 **terraform**을 사용해보고 싶습니다! 코드로 인프라를 관리하다 보니까 서로 인프라 관련된 공유나 학습에 관해서도 보다 편할 것 같아요!

플젝하다보면 VM 옮기고 이런 상황 몇번 생길거 같아서 적극 찬성입니다!

**Reply by belljun3395** [reply]
## 인스턴스
<img width="701" alt="스크린샷 2024-05-26 오후 8 33 36" src="https://github.com/YAPP-Github/24th-Web-Team-1-BE/assets/102807742/1079339a-3e94-45e1-b574-fa4aa8a70e4a">

## DB
<img width="1209" alt="스크린샷 2024-05-26 오후 8 39 49" src="https://github.com/YAPP-Github/24th-Web-Team-1-BE/assets/102807742/d13d2921-42b4-4773-a7c9-d2a30ef7a564">

NCP를 사용하면 기본 서버비가 3.4만 X 2 (서버 + 프런트) 해서 6.8만원 나올꺼 같아요
여기에 DB가 한 10만원 정도 나올꺼 같고
Public Ip, LoadBalancer, Object Storage 이런거 더 하면 한 20만원 언저리 나올꺼 같은데요...???

이정도면 얍 지원 끊기면 바로 장례식 갈꺼 같은데요...?

**Reply by belljun3395** [reply]
NCP는 AI 사용 하는 쪽으로 사용하고 AWS 프리티어에 맞추는 것도 좋을 것 같다는 생각이 듭니다..?!

**Reply by belljun3395** [reply]
추가로 현재 필요한 인프라는 아래와 같을꺼 같아요!

## 꼭 필요한 인프라

- 서버 인스턴스 1대
- 서버 DB
- 서버 Object Store
- 클라이언트 인스턴스 1대

## 배치성 인프라

- GPT 문제 생성 인스턴스 1대 
- 이메일 전송 인스턴스

**Reply by hun-ca** [reply]
일단은 DB + was 로만 가고 나머지는 싹다 was 서바애 두고 was scale up해가다가 안되겠다 싶을때 이메일 서버 이런거 빼죠

**Reply by belljun3395** [reply]
알겠슴다~ 
근데 GPT 셀리리움 요것도 뺄 수 있나요?

**Reply by belljun3395** [reply]
추가로 was 서버에 싹다 두는 대신 이메일 관련된 부분은 멀티 모듈로 따로 구현하는 것은 어때요?

**Reply by hun-ca** [reply]
ㅋㅋㅋㅋㅋ 아 셀레니움.. 할까요? 그거 API는 가격 얼마에요?

**Reply by belljun3395** [reply]
배치도 서버 안두는 거면 스케쥴러 말고 람다 같은걸로 하면 어떨까요?

**Reply by belljun3395** [reply]
API는 gpt 말하는거죠?

**Reply by hun-ca** [reply]
네네 그 유료 GPT API요
근데 이거 클로버에도 있지 않나?

**Reply by belljun3395** [reply]
근데 GPT랑 쫌 다른 느낌인거 같아요..!
완전 챗봇을 위한..?

**Reply by belljun3395** [reply]
저도 GPT 초짜라 다시 계산해봐야겠지만 제가 데모로 구현한 것처럼 유저가 GPT를 사용하게 하는게 아니라면 어느정도 조절 할 수 있을 것 같아요!

우선 지금은 API 콜 하나당 0.01달러 정도 나오는거 같슴다..
파인튜닝 시키고 입려도 많이해서 쫌 가격이 나가는거 같아요

**Reply by hun-ca** [reply]
100번 콜 -> 약 1500언?

**Reply by hun-ca** [reply]
할만 한데

**Reply by belljun3395** [reply]
근데 또 문제 재대로 뽑을라면 파인튜닝 해야하는데 그거하면 돈이 팍팍 깍입니다..ㅎㅎ

**Reply by belljun3395** [reply]
근데 막 10만원 이렇게 나오지는 않을꺼 같아요..? 아마..?!

**Reply by belljun3395** [reply]
뭐 정 안되면 영모가 고생을...ㅋㅋ

**Reply by hun-ca** [reply]
100번 콜 1500원에 우리 6명
1500/6 = 250원
100번 콜에 인당 250원

**Reply by hun-ca** [reply]
셀레니움도 함 도전해보겠음

**Reply by belljun3395** [reply]
NCP 쓴다. 
=> 지원 항목 확인해보기

NCP에서 제공하는 DEV TOOL 사용해도 나쁘지 않을 것 같다.

배치 관련된 것은 따로 서버를 두지는 않는다.
배치를 API로 구현해야된다.
Git Action Cron을 활용하는 방법으로!

**Reply by belljun3395** [reply]
==========================================================

**Reply by belljun3395** [reply]
S3와 같은 오브젝트 저장소가 필요할 것 같습니다

**Comment by belljun3395** [comment]
# DB 선택

현 상황에서는 CUD보다는 R이 압도적으로 많은 상황

RDB와 NoSQL 사이 선택이 필요

**Reply by belljun3395** [reply]
저도 고민을 해봤는데 MySQL에 Json 타입을 지원하는데 그것을 어떻게 생각하시나요?

**Reply by belljun3395** [reply]
FK는 안잡는다.

Entity 관계를 위해서는 
1. 문서를 잘 적는다.
2. 상위 클래스로 만든다.

**Reply by belljun3395** [reply]
==========================================================

**Reply by belljun3395** [reply]
RDB / NoSQL 결정 필요 -> RDB일 경우 JPA 벤, jooq 적극고려

**Comment by belljun3395** [comment]
# 소스코드 아키텍처

패키지 구조, 코드 컨벤션과 같은 개발 이전에 논의 할 사항 존재

**Reply by belljun3395** [reply]
참고 레포 
- 지훈 : https://github.com/ok-su-su/oksusu-susu-api
- 종준 : https://github.com/depromeet12th/three-days-server

**Reply by belljun3395** [reply]
https://github.com/ok-su-su/oksusu-susu-api

요기 요약 정리하면 

동일 도메인 사용 서비스 클래스 : XXXService
다른 도메인 함께 사용 서비스 클래스 XXXFacadeService
컨트롤러 : Resource

요정도 될까요?

**Reply by belljun3395** [reply]
도메인 모델 도입한다. => 엔티티 모델에는 메서드가 없다

Usecase.execute 패턴을 도입한다.

Repository 조회할 때는 Query객체를 사용한다.

**Comment by belljun3395** [comment]
# 스프링, 자바, 코틀링

버전 선택 필요

언어 선택 필요

라이브러리 선택 필요

멀티 모듈 여부 선택

**Reply by belljun3395** [reply]
웹플럭스 Vs MVC 결정

웹플럭스를 사용한다면 R2DBC를 사용해야 함

블록/논블록킹에 주의하며 코드를 작성해하 함

**Reply by hun-ca** [reply]
R2DBC 안써봐서 러닝 코스트 좀 있을듯... 고민좀 해볼게요

**Reply by belljun3395** [reply]
- 코루틴, webflux 공부하고 결정하기

**Comment by belljun3395** [comment]
# CICD

깃헙 액션을 통한 CD 구축

테스트 코드는 얼마나 작성할지

**Reply by hun-ca** [reply]
CI 서버 따로 구축하기 어려운 상황이라 깃헙 액션 사용 적극 찬성입니다

**Comment by belljun3395** [comment]
# Git

브렌치 관리

커밋 컨벤션

PR

**Comment by belljun3395** [comment]
저는 개인 플젝을 할 때는 도메인 객체를 도입해서 비즈니스 로직을 처리하는데 혹시 도메인 객체에 대한 어떤 생각을 가지고 계신지 궁금합니다.

저는 아래와 같습니다.
- 엔티티 객체: DB 테이블과 대응되는 필드만 가지고 있어야 한다.
- 도메인 객체: 요청에서 필요한 정보만 가지고 있고 요청의 비즈니스 로직을 수행합니다.

조금 부끄럽지만 관련 생각 정리한 블로그 [링크](https://belljundev.tistory.com/15)도 첨부합니다~~!

**Comment by belljun3395** [comment]
이메일 전송 관련 부분 클라이언트와 기능 구현에 관해서 이야기가 필요할 것 같은데 현업에서는 이메일 전송 부분은 어떤 파트 담당일까요?

보내는 것만 생각하면 서버에서 해야할 것 같은데 안에 내용을 생각하면 클라의 도움이 필요할 것 같기도 하고..

**Reply by belljun3395** [reply]
이메일 관련 부분은 전송은 서버, 템블릿은 클라로!

**Comment by belljun3395** [comment]
## 멀티 모듈

- common
- api ( common, data, email ) - ApiMain.class
- data ( entity )
- email ( common )

### 추후 VM 할당
- batch ( email ) - BatchMain.class

**Comment by belljun3395** [comment]
- 플젝 초기 셋업
  - CICD 작업
  - 로직개발

**Reply by belljun3395** [reply]
플젝 초기 셋업 
: 소스코드 레벨 예제 코드가 필요

CICD
: @belljun3395 담당 but @hun-ca 에게 컨펌
NCP

// MVP가 정해지고난 이후
API 명세 => rest docs / controller test (ok)

DB 설계

로직 개발
